DBZ-1082 Add new custom recovery mode, more metadata

addisonj opened the pull request:
This commit does a few things:

- Refactors snapshot modes to be encapsulated by an interface and
to only use that interface in determining when to snapshot and in
determing the type of the `RecordProducer` interface to instantiate
- Refactors the configuration of existing snapshot modes to tie the
existing snapshot modes to their aligned implementation
- Adds a new snapshot.mode, custom, and a new configuration option to
specify a custom implementation that will be loaded by the class loader
- Changes the visibility of some classes to allow for custom snapshot
modes to get enough context to make an informed choice
- Adds some metadata about slots (the catalog_xmin) to give a full idea
of the state of slots which can be useful in implementing snapshot
modes (which is also configurable, as it can add some overhead)

Together, these changes allow for a much broader flexibility got end
users to implement a snapshot mode that can do more advanced snapshots,
such as partial recovery or for partial snapshots for tables where not
all records are needed.

This could also be seen as superseeding the
`snapshot.select.statement.overrides` to allow for users to dynamically
build queries based on the state of the slot and the offsets consumed.

addisonj commented:
Continuing on from #730, this is a a not-yet-finished idea of what implementing custom snapshot modes would look like.

Would love to get your feedback @gunnarmorling. Here are a few of still open questions for me:

- Any concerns with changing the visibility of some of these classes and methods? They are limited and pretty much just getters, so it seems sane to me
- Are the names clear enough? In the prior version I had called it `SnapshotMode`, but that has a conflict with the enum (which could be renamed), but this seemed good enough to get the point accross
- Should we do something with a different class loader? IDK if anyone is really going to need to use a custom class loader, but that could *potentially* be important in some environments, but I am not as familiar with concerns there.

Once again, this isn't 100% complete yet, I want to write an integration test with a custom snapshot mode and also exercise the functionality of tracking xmins, as well as fix a few merge conflicts, but I wanted to get a discussion around this started.

Thanks in advance!

addisonj commented:
This has tests now so it should be ready for some feedback :smile: thanks!

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresTaskContext.java:
```diff
@@ -25,10 +28,15 @@
     private final TopicSelector<TableId> topicSelector;
     private final PostgresSchema schema;
 
+    private Long lastXminTime;
```
Can we make this an `Instant` or something similar? We're trying to converge to the more semantic JDK 8 temporal data types where possible.

addisonj replied:
done

addisonj commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:
```diff
@@ -95,34 +98,54 @@ public static HStoreHandlingMode parse(String value,String defaultValue){
+        CUSTOM("custom", (s) -> {
+            Snapshotter sn = null;
+            try {
+                Class clazz = Class.forName(s, true, ClassLoader.getSystemClassLoader());
```
oh nifty, that is a lot nicer

addisonj commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorTask.java:
```diff
@@ -80,42 +83,26 @@ public void start(Configuration config) {
             try (PostgresConnection connection = taskContext.createConnection()) {
                 logger.info(connection.serverInfo().toString());
+                slotInfo = connection.fetchReplicationSlotInfo(connectorConfig.slotName(), connectorConfig.plugin().getPostgresPluginName());
+            } catch (SQLException e) {
```
oh yes, sorry, I had adjust my IDE settings to do this and then must have lost them when I made some other changes

addisonj commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:
```diff
@@ -95,34 +98,54 @@ public static HStoreHandlingMode parse(String value,String defaultValue){
+        private final Function<String, Snapshotter> builderFunc;
 
-        SnapshotMode(String value) {
+        SnapshotMode(String value, Function<String, Snapshotter> buildSnapshotter) {
```
done

addisonj commented:
@gunnarmorling yeah, it felt *slightly* weird, but mostly what got exposed was read only methods.

I just pushed another commit that introduce a new package with some very simple public types that expose everything I think the user could really want and it is all stuff that is very inherent to how PG logical decoding works (lsn, txId, xmin, etc)

I am not sure if there is a more elegant way to do that, but it is fairly simple and causes a low amount of overhead.

This also makes it easier for third parties to directly test the logic of their snapshotter, so that is nice too!

Thanks for the feedback and let me know if that addresses your concerns!

addisonj commented:
just a small bump on this @gunnarmorling, if you have any feedback, I would appreciate it. I have some time in the next few days to work on this and then I will be getting a bit busy with other things

jpechane commented:
@addisonj Hi, Gunnar is on holidays so I am now reviewing instead of him, unfortunately it is a fresh start for me so the feedback will be a little bit delayed.

jpechane commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsSnapshotProducer.java:
```diff
@@ -207,26 +208,30 @@ private void takeSnapshot(BlockingConsumer<ChangeEvent> consumer) {
-                            tableId, Strings.duration(clock().currentTimeInMillis() - exportStart));
-                    rowsCounter.set(0);
+                    final String selectStatement = snapshotter.buildSnapshotQuery(tableId);
+                    if (selectStatement == null || selectStatement.isEmpty()) {
```
If we support omitting of query for table I thing we should use `Optional` insted of `null`

addisonj replied:
yeah, I wasn't totally sure on this... it seemed like a sane thing to do, just for a defensive programming for a user controlled class, but changing the signature communicates that more if that feels more like "feature" rather than an implementation detail that prevents a NPE.

addisonj commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:
```diff
@@ -750,6 +791,18 @@ public static SchemaRefreshMode parse(String value) {
+            .withDisplayName("Xmin fetch interval (ms)")
+            .withType(Type.LONG)
+            .withWidth(Width.SHORT)
+            .withDefault(-1L)
```
Zero seems like it might not be as intuitive, if an interval is zero, I would expect (and perhaps want?) to always be fetching the latest value. In reality, 1 ms is probably just as good as zero for that case, but I would be more worried about someone setting zero and not realizing that it meant disabling the feature.

addisonj replied:
that said, if you still prefer zero (not sure if there is prior art for this) I am happy to change this, just let me know

jpechane replied:
`heartbeat.interval.ms` and `snapshot.delay.ms` already use `0` or (value `<=0`) as such so I think we should be consistent

addisonj commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresTaskContext.java:
```diff
@@ -56,6 +68,37 @@ protected void refreshSchema(boolean printReplicaIdentityInfo) throws SQLExcepti
+            return null;
+        }
+
+        if (lastXminTime == null || getClock().currentTimeAsInstant().isAfter(lastXminTime.plus(xminFetchTime))) {
```
looks like it should work, will do

addisonj commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorTask.java:
```diff
@@ -70,6 +71,12 @@ public void start(Configuration config) {
+        Snapshotter snapshotter = connectorConfig.getSnapshotter();
+        if (snapshotter == null) {
+            logger.error("Unable to load snapshotter, if using custom snapshot mode, double check your settings");
+            return;
```
will do, I think I came across some other piece of code that did something similar, but I like a runtime exception more here as well

addisonj commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:
```diff
@@ -750,6 +791,18 @@ public static SchemaRefreshMode parse(String value) {
+                    "This xmin value is exposed by the slot which gives a lower bound of where a new replication slot could start from. " +
+                    "The lower the value, the more likely this value is to be the current 'true' value, but the bigger the performance cost. " +
+                    "The bigger the value, the less likely this value is to be tue current 'true' value, but the lower the performance penalty. " +
+                    "The default is set to -1 ms, which tracking xmin.");
```
yep, will fix

jpechane commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsSnapshotProducer.java:
```diff
@@ -62,14 +61,16 @@
         currentRecord = new AtomicReference<>();
-        if (continueStreamingAfterCompletion) {
+        this.snapshotter = snapshotter;
+        if (snapshotter.recordProducerType() == Snapshotter.RecordProducerType.SNAPSHOT_STREAM) {
```
Yeah, I'd prefer that, I think it is easiser to use and comprehend instead of 3-state enum

jpechane commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/types/OffsetState.java:
```diff
@@ -0,0 +1,46 @@
+ * A simple data container that represents the last seen offset
+ * which was written by debezium.
+ */
+public class OffsetState {
```
Not as snapshotters but as `Snapshotter` interface.

jpechane commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/snapshot/Snapshotter.java:
```diff
@@ -0,0 +1,26 @@
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.postgresql.snapshot;
```
spi - service provided interface. Yes, if possible I'd like to have classes/interfaces consumable by an extrenal code in an isolated package as a well-defined contract.

gunnarmorling replied:
@addisonj, SPI means a special kind of API geared towards extensions and things like that, i.e. making certain aspects of the core library customizeable and pluggable. So it's not targettin mere *users* of an API but *implementors/customizers*.

addisonj commented:
@jpechane made all the requested changes, please let me know if there is anything else!

jpechane commented:
@gunnarmorling This PR is good to go from my side (when conflicts are resolved)
@addisonj I apologize but there was another PR that was convertting timestmps from long to Instant. It cause versy small conflict in this PR and moreover `Clock` now contains methods very similar to those you've introduced. If you could rebase it to the lastest master we can merge the PR then. Thanks for the great work.

gunnarmorling commented:
Thanks for the reminder, @jpechane!

Hey @addisonj, I'm back from holidays, will take a look at this in the course of this week hopefully. Thanks a lot for this work and thanks for you patience.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/spi/Snapshotter.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.Optional;
+
+public interface Snapshotter {
+    void init(PostgresConnectorConfig config, OffsetState sourceInfo,
```
What is the exact info needed from the given config?

addisonj replied:
Currently, that is used for the built-in snapshotters so they can still respect the snapshotSelectOverrides. It still seems like some config values might be useful to be referenced by custom snapshotters as well though, such as doing things with the table whitelist or some other config values.

gunnarmorling replied:
ok.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/spi/Snapshotter.java:
```diff
@@ -0,0 +1,21 @@
+    boolean shouldSnapshot();
+    boolean shouldStream();
+
+    Optional<String> buildSnapshotQuery(TableId tableId);
```
Do implementations need the database name actually? If not, I'd suggest to simply pass in table and schema name as two strings, thus avoiding the reference to `TableId`.

addisonj replied:
The built-in snapshotters (via the `QueryingSnapshotter`) make use of the `TableId` class as that has utility methods for escaping the table names, etc. While it is exposing one more class, the TableId class is quite useful and has no internal concerns (AFAICT) associated with it

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/spi/SlotState.java:
```diff
@@ -0,0 +1,40 @@
+ * A simple data container that holds the state of the current slot
+ */
+public class SlotState {
+    private final Long latestFlushedLSN;
```
Minor nit: I think we used `Lsn` casing in other places (applies in some other parts of the PR, too).

addisonj replied:
I looked around, and both casings seems to be used. For example, all local variables in SlotState use this casing (`LSN`) as well as most of `PostgresConnection`.

I am happy to standardize on one of the other, but (AFAICT) it seems like most local variables use the all upcase version and methods use the `Lsn` version. Let me know which you prefer

gunnarmorling replied:
Let's use `Lsn` consistently. It's the style we generally prefer (as you say, there still may be existing usages of the other one).

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:
```diff
@@ -750,6 +791,19 @@ public static SchemaRefreshMode parse(String value) {
                     "have TOASTed data that are rarely part of these updates. However, it is possible for the in-memory schema to " +
                     "become outdated if TOASTable columns are dropped from the table.");
+
+    public static final Field XMIN_FETCH_INTERVAL = Field.create("xmin.fetch.interval")
```
Can you check with other similar options, I think we suffixed them with ".ms".

addisonj replied:
will change

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/spi/OffsetState.java:
```diff
@@ -0,0 +1,46 @@
+        return xmin;
+    }
+
+    public Long lastCommitTs() {
```
Should be rather `Instant` instead?

addisonj replied:
will change

addisonj commented:
@gunnarmorling hope your time off was nice, I was gone last week as well, so just getting to this!

I updated the PR, but still just a few outstanding questions, so let me know where you are on these and will make the changes. Glad to hear we can get it in the next release.

As far as other changes I have, this is the last thing I have on my side that I know about right now, however, there are a few things that I am concerned about as we start using this more. (These concerns are outside the scope of this issue, but I am not sure where it makes sense to have this conversations)
1. I am curious about more aggressive snapshotting, i.e., doing multiple tables in parallel. This seems like for large databases with lots of tables, this could be something really useful
2. Metrics, I went and looked at all of the built-in metrics abstractions and noticed that they should be more automatic as the postgres connector moves to the newer style that SqlServer is doing, but I am not sure of the timeline for that refactor. if it is soon, then we can wait, otherwise, I was considering doing a small change to add at least some metrics until that refactor can take place.

Would love your thoughts and thanks for the review on this one :)

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/SourceInfo.java:
```diff
@@ -99,6 +101,7 @@
                                                      .field(TIMESTAMP_KEY, Schema.OPTIONAL_INT64_SCHEMA)
                                                      .field(TXID_KEY, Schema.OPTIONAL_INT64_SCHEMA)
                                                      .field(LSN_KEY, Schema.OPTIONAL_INT64_SCHEMA)
+                                                     .field(XMIN_KEY, Schema.OPTIONAL_INT64_SCHEMA)
```
This probably needs to go to the end of the block, so to avoid backwards-compatibility issues when using Avro and the schema registry. @jpechane, can you confirm as per your exploration on the other PR?

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/ServerInfo.java:
```diff
@@ -156,14 +162,26 @@ protected boolean active() {
         }
+
+        protected Long restartLSN() {
+            return restartLSN;
```
Is there a JavaDoc of the semantics of "restartLsn" vs. "latestFlushedLsn" somewhere? Doesn't have to be repeated in every field/getter representing it, but should be there once, in the "most natural" place.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/spi/Snapshotter.java:
```diff
@@ -0,0 +1,21 @@
+
+import java.util.Optional;
+
+public interface Snapshotter {
```
Can you add a class-level comment? It's a good idea for all the types in the SPI package so implementors understand about their semantics.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/spi/OffsetState.java:
```diff
@@ -0,0 +1,48 @@
+ * A simple data container that represents the last seen offset
+ * which was written by debezium.
+ */
+public class OffsetState {
```
Can you add a new annotation `io.debezium.annotation.Incubating` (in _debezium-core_) for marking incubating APIs/SPIs and put it to all the new types in this SPI package.

gunnarmorling replied:
This one can serve as example: https://github.com/hibernate/hibernate-validator/blob/master/engine/src/main/java/org/hibernate/validator/Incubating.java

gunnarmorling commented:
> I am curious about more aggressive snapshotting, i.e., doing multiple tables in parallel. This seems like for large databases with lots of tables, this could be something really useful

Yes, totally agree. This is on the roadmap (see [DBZ-823](https://issues.jboss.org/browse/DBZ-823)). It makes sense to implement this once the PG connector has been "migrated" to some framework infrastructure we've implemented for the Oracle and SQL Server connectors ([DBZ-777](https://issues.jboss.org/browse/DBZ-777))). This will allow us to do this work just once. @jpechane is currently working on that migration task.

> Metrics, I went and looked at all of the built-in metrics abstractions and noticed that they should be more automatic as the postgres connector moves to the newer style that SqlServer is doing, but I am not sure of the timeline for that refactor. if it is soon, then we can wait, otherwise, I was considering doing a small change to add at least some metrics until that refactor can take place.

As you say, we'd also get this "for free" for the PG connector once DBZ-777 has been resolved. The target release for this is 0.10. So I don't think there's much value in adding "proprietary" metrics to the PG connector at this point, as this effort would be largely wasted and obsolete soon.

gunnarmorling commented:
Could you also rebase to the current master? I don't think the conflict is hard to resolve.

addisonj commented:
@gunnarmorling okay! I think I addressed most of those changes, let me know if there is anything else

gunnarmorling commented:
Rebased and applied. Thank you so much, @addisonj! One last thing: can you file a PR for the docs repo, describing the new option and also touch on the new SPI type?

Perhaps it's even worth a small dedicated blog post describing the possible usage of it. WDYT, would you be interested in that?
